# NRC Emotional Lexicon

This is the [NRC Emotional Lexicon](http://saifmohammad.com/WebPages/NRC-Emotion-Lexicon.htm): "The NRC Emotion Lexicon is a list of English words and their associations with eight basic emotions (anger, fear, anticipation, trust, surprise, sadness, joy, and disgust) and two sentiments (negative and positive). The annotations were manually done by crowdsourcing."

I don't trust it, but everyone uses it.

<p class="reading-options">
  <a class="btn" href="/upshot-trump-emolex/nrc-emotional-lexicon">
    <i class="fa fa-sm fa-book"></i>
    Read online
  </a>
  <a class="btn" href="/upshot-trump-emolex/notebooks/NRC Emotional Lexicon.ipynb">
    <i class="fa fa-sm fa-download"></i>
    Download notebook
  </a>
  <a class="btn" href="https://colab.research.google.com/github/littlecolumns/ds4j-notebooks/blob/master/upshot-trump-emolex/notebooks/NRC Emotional Lexicon.ipynb" target="_new">
    <i class="fa fa-sm fa-laptop"></i>
    Interactive version
  </a>
</p>

### Prep work: Downloading necessary files
Before we get started, we need to download all of the data we'll be using.
* **NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt:** NRC Emotional Lexicon - a list of English words and their associations with eight basic emotions and two sentiments


In [1]:
# Make data directory if it doesn't exist
!mkdir -p data
!wget -nc https://nyc3.digitaloceanspaces.com/ml-files-distro/v1/upshot-trump-emolex/data/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt -P data

--2021-11-24 15:39:16--  https://nyc3.digitaloceanspaces.com/ml-files-distro/v1/upshot-trump-emolex/data/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt
Resolving nyc3.digitaloceanspaces.com (nyc3.digitaloceanspaces.com)... 162.243.189.2
Connecting to nyc3.digitaloceanspaces.com (nyc3.digitaloceanspaces.com)|162.243.189.2|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2581050 (2.5M) [text/plain]
Saving to: ‘data/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt’

NRC-emotion-lexicon 100%[===================>]   2.46M  2.89MB/s    in 0.9s    

2021-11-24 15:39:17 (2.89 MB/s) - ‘data/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt’ saved [2581050/2581050]



In [3]:
!pip3 install pandas

DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
     |████████████████████████████████| 4.7 MB 9.3 MB/s            
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     |████████████████████████████████| 12.4 MB 21.9 MB/s            
  Using cached pytz-2021.3-py2.py3-none-any.whl (503 kB)
  Created wheel for pandas: filename=pandas-1.3.4-cp39-cp39-macosx_11_0_arm64.whl size=10236252 sha256=0544f7720a06ca97d85904a98f2c335038c3f75c4a4f2ea1f8382676db85b4c3
  Stored in directory: /Users/pratikaher/Library/Caches/pip/wheels/44/1a/bb/e1d1c2bc72adfa0c96d573d26755d412caa9a1a60b6898983c
Successfully built pandas
  DEPRECATION: Configuring installation scheme with distutils config files is deprecated 

In [4]:
import pandas as pd

In [5]:
filepath = "data/NRC-Emotion-Lexicon-v0.92/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt"
emolex_df = pd.read_csv(filepath,  names=["word", "emotion", "association"], skiprows=45, sep='\t', keep_default_na=False)
emolex_df.head(12)

FileNotFoundError: [Errno 2] No such file or directory: 'data/NRC-Emotion-Lexicon-v0.92/NRC-emotion-lexicon-wordlevel-alphabetized-v0.92.txt'

Seems kind of simple. A column for a word, a column for an emotion, and whether it's associated or not. You see "aback aback aback aback" because there's a row for every word-emotion pair.

## What emotions are covered?

Let's look at the 'emotion' column. What can we talk about?

In [27]:
emolex_df.emotion.unique()

array(['anger', 'anticipation', 'disgust', 'fear', 'joy', 'negative',
       'positive', 'sadness', 'surprise', 'trust'], dtype=object)

In [28]:
emolex_df.emotion.value_counts()

negative        14182
trust           14182
anger           14182
anticipation    14182
sadness         14182
fear            14182
joy             14182
disgust         14182
positive        14182
surprise        14182
Name: emotion, dtype: int64

## How many words does each emotion have?

Each emotion doesn't have 14182 words associated with it, unfortunately! `1` means "is associated" and `0` means "is not associated."

We're only going to care about "is associated."

In [29]:
emolex_df[emolex_df.association == 1].emotion.value_counts()

negative        3324
positive        2312
fear            1476
anger           1247
trust           1231
sadness         1191
disgust         1058
anticipation     839
joy              689
surprise         534
Name: emotion, dtype: int64

In theory things could be *kind of* angry or *kind of* joyous, but it doesn't work like that. If you want to spend a few hundred dollars on Mechnical Turk, though, *your own personal version can.*

## What if I just want the angry words?

In [30]:
emolex_df[(emolex_df.association == 1) & (emolex_df.emotion == 'anger')].word

30          abandoned
40        abandonment
170             abhor
180         abhorrent
270           abolish
             ...     
141220       wrongful
141230        wrongly
141470           yell
141500           yelp
141640          youth
Name: word, Length: 1247, dtype: object

## Reshaping

You can also reshape the data in order to look at it a slightly different way

In [31]:
emolex_words = emolex_df.pivot(index='word', columns='emotion', values='association').reset_index()
emolex_words.head()

emotion         word  anger  anticipation  disgust  fear  joy  negative  \
0              aback      0             0        0     0    0         0   
1             abacus      0             0        0     0    0         0   
2            abandon      0             0        0     1    0         1   
3          abandoned      1             0        0     1    0         1   
4        abandonment      1             0        0     1    0         1   

emotion  positive  sadness  surprise  trust  
0               0        0         0      0  
1               0        0         0      1  
2               0        1         0      0  
3               0        1         0      0  
4               0        1         1      0

You can now pull out individual words...

In [32]:
# If you didn't reset_index you could do this more easily
# by doing emolex_words.loc['charitable']
emolex_words[emolex_words.word == 'charitable']

emotion        word  anger  anticipation  disgust  fear  joy  negative  \
2001     charitable      0             1        0     0    1         0   

emotion  positive  sadness  surprise  trust  
2001            1        0         0      1

...or individual emotions....

In [33]:
emolex_words[emolex_words.anger == 1].head()

emotion         word  anger  anticipation  disgust  fear  joy  negative  \
3          abandoned      1             0        0     1    0         1   
4        abandonment      1             0        0     1    0         1   
17             abhor      1             0        1     1    0         1   
18         abhorrent      1             0        1     1    0         1   
27           abolish      1             0        0     0    0         1   

emotion  positive  sadness  surprise  trust  
3               0        1         0      0  
4               0        1         1      0  
17              0        0         0      0  
18              0        0         0      0  
27              0        0         0      0

...or multiple emotions!

In [34]:
emolex_words[(emolex_words.joy == 1) & (emolex_words.negative == 1)].head()

emotion        word  anger  anticipation  disgust  fear  joy  negative  \
61        abundance      0             1        1     0    1         1   
1018           balm      0             1        0     0    1         1   
1382     boisterous      1             1        0     0    1         1   
1916      celebrity      1             1        1     0    1         1   
2004        charmed      0             0        0     0    1         1   

emotion  positive  sadness  surprise  trust  
61              1        0         0      1  
1018            1        0         0      0  
1382            1        0         0      0  
1916            1        0         1      1  
2004            1        0         0      0

The useful part is going to be just getting words for a **single emotion.**

In [35]:
# Angry words
emolex_words[emolex_words.anger == 1].word

3          abandoned
4        abandonment
17             abhor
18         abhorrent
27           abolish
            ...     
14122       wrongful
14123        wrongly
14147           yell
14150           yelp
14164          youth
Name: word, Length: 1247, dtype: object

## Review

We took a quick look at the **Emotional Lexicon**, a sentiment analysis library that includes multiple emotional axes instead of just "positive" and "negative."

## Discussion topics

The Emotional Lexicon used words tagged individually by internet users. Do you think this is an effective method for understanding sentiment?

How does this method compare to the [Sentiment140](http://www.sentiment140.com/) method that we covered in sentiment analysis?